apps/omg_watcher_info/test/omg_watcher_info/pending_block_processor_test.exs

```diff
@@ -0,0 +1,199 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.WatcherInfo.PendingBlockProcessorTest do
+  use ExUnitFixtures
+  use ExUnit.Case, async: true
+
+  import Ecto.Query, only: [from: 2]
+  import OMG.WatcherInfo.Factory
+
+  alias OMG.WatcherInfo.DB
+  alias OMG.WatcherInfo.DB.PendingBlock
+  alias OMG.WatcherInfo.PendingBlockProcessor
+  alias OMG.WatcherInfo.PendingBlockProcessor.Storage
+
+  @interval 200
+
+  defmodule DumbStorage do
+    use GenServer
+
+    def start_link(args) do
+      GenServer.start_link(__MODULE__, args, name: __MODULE__)
+    end
+
+    def init(args) do
+      # note: each mock will set the state of the corresponding
+      # function name to {count, metadata} where count represents the amount of time
+      # the function was called and metadata is a map of additional info.
+      default_state = %{
+        get_next_pending_block: {0, nil},
+        process_block: {0, nil},
+        next_pending_block: nil
+      }
+
+      given_state = Enum.into(args, %{})
+      state = Map.merge(default_state, given_state)
+      {:ok, state}
+    end
+
+    def get_next_pending_block() do
+      GenServer.call(__MODULE__, :get_next_pending_block)
+    end
+
+    def process_block(block) do
+      GenServer.call(__MODULE__, {:process_block, block})
+    end
+
+    def handle_call(:get_next_pending_block, _from, %{get_next_pending_block: {count, _}} = state) do
+      {:reply, state.next_pending_block, %{state | get_next_pending_block: {count + 1, state.next_pending_block}}}
+    end
+
+    def handle_call({:process_block, block}, _from, %{process_block: {count, _}} = state) do
+      {:reply, {:ok, nil}, %{state | process_block: {count + 1, block}}}
+    end
+  end
+
+  defp setup_with_storage(storage, tags) do
+    {:ok, pid} =
+      PendingBlockProcessor.start_link(
+        processing_interval: @interval,
+        storage_module: storage,
+        name: PendingBlockProcessorTest
+      )
+
+    Map.put(tags, :pid, pid)
+  end
+
+  describe "handle_info/2 with mocked Storage" do
+    setup tags do
+      setup_with_storage(DumbStorage, tags)
+    end
+
+    test "calls get_next_pending_block/0 when triggered", %{pid: pid} do
+      storage_pid = start_storage_mock()
```

Comment:
hm... I don't understand how this works.
you're calling `start_storage_mock/0` in every test, but this process is a named process:
 `GenServer.start_link(__MODULE__, args, name: __MODULE__)`.

The internal state of this module is being updated and accessed from all tests (except the ones below). So the only explanation is that these are not executed async?

Reply:
okay, so the reason is that everything inside a describe is executed top-down.
which wasn't obvious to me.
here's a couple of suggestions:
- move the DumbStorage to the bottom
- move setup_with_storage/0 to the bottom

couple DumpStorage with the instance of PendingBlockProcessor with an Agent. Agent holds key pairs %{pending block processor pid => dump storage pid}. then you can drop the DumpStorage naming. 

To me, it's easier to reason about your tests. I didn't understand why all tests start new PendingBlockProcessor but they all funnel towards the same instance of storage. 

Also, it seems you could have accomplished the same thing with Postgres Sandbox?

Reply:
Will remove the mocked storage, the genserver is too simple now to require mocking.